pkg/chunk/object/chunk_object_client.go

```diff
@@ -0,0 +1,80 @@
+package object
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/util"
+)
+
+// ChunkClient is used to store chunks in object store backends
+type ChunkClient struct {
+	Store chunk.ObjectClient
+}
+
+// Stop shuts down the object store and any underlying clients
+func (o *ChunkClient) Stop() {
+	o.Store.Stop()
+}
+
+// PutChunks stores the provided chunks in the configured backend. If multiple errors are
+// returned, the last one sequentially will be propagated up.
+func (o *ChunkClient) PutChunks(ctx context.Context, chunks []chunk.Chunk) error {
```

Comment:
This is not an exact drop-in replacement. This is the copy of the S3 client, while other clients (GCS, FS, Azure) didn't have any parallelization. What may be the implications of this change?

Reply:
I've thought about this a bit for an unrelated reason. The short answer is it doesn't matter if we parallelize writes here because we never pass more the one chunk at a time to the client:

From: pkg/chunk/series_store.go
```go
// Put implements ChunkStore
func (c *seriesStore) Put(ctx context.Context, chunks []Chunk) error {
	for _, chunk := range chunks {
		if err := c.PutOne(ctx, chunk.From, chunk.Through, chunk); err != nil {
			return err
		}
	}
	return nil
}
```

From: pkg/chunk/chunk_store.go
```go
// Put implements ChunkStore
func (c *store) Put(ctx context.Context, chunks []Chunk) error {
	for _, chunk := range chunks {
		if err := c.PutOne(ctx, chunk.From, chunk.Through, chunk); err != nil {
			return err
		}
	}
	return nil
}

// PutOne implements ChunkStore
func (c *store) PutOne(ctx context.Context, from, through model.Time, chunk Chunk) error {
	chunks := []Chunk{chunk}

	err := c.storage.PutChunks(ctx, chunks)
	if err != nil {
		return err
	}

	c.writeBackCache(ctx, chunks)

	writeReqs, err := c.calculateIndexEntries(chunk.UserID, from, through, chunk)
	if err != nil {
		return err
	}

	return c.index.BatchWrite(ctx, writeReqs)
}
```

So in general, writing in parallel on the client level is irrelevant. I chose the S3 implementation because I felt it was the most advanced implementation. If we want to keep a sequential implementation, the client could be updated to use a maximum number of concurrent writes, and set that value to one. I think in general we should support parallel/batch writes, which is why I chose to use the same approach as the S3 client. However, it does not matter at this point in time because of how the clients are used upstream.

Reply:
You're right. I double checked it, and we always call `PutChunks()` with 1 chunk. I'm OK keeping parallelism then. I'm not against parallelism, but I was wondering potential side effects (ie. rate limiting) of this change, but given there's no practical change LGTM.